### PR TITLE
fix: XCom state logic

### PR DESCRIPTION
**Pull request by Mryashbhardwaj**

### CODE REVIEW

File: ext/scheduler/airflow/__lib.py [hunk 1]
```diff
@@ -495,20 +497,28 @@ def get_result_for_monitoring_from_xcom(ctx):
...
         ti = ctx.get('task_instance')
-        key = "sensorEvt/{}/{}/{}".format(ti.task_id , ctx.get('next_execution_date').strftime(TIMESTAMP_FORMAT) , ti.try_number)
-        session.query(XCom).filter(XCom.key == key).delete()
+        cleanup_key_patttern = "sensorEvt/{}".format(ti.task_id)
+        # cleanup xcom is runs after the task run is completed hence the try number is increased by the time cleanup_xcom is called
+        log.info(f'xcom key pattern to clear -> {cleanup_key_patttern}')
+
+        result = session.query(XCom).filter( and_( XCom.dag_id == ti.dag_id , XCom.key.startswith(cleanup_key_patttern))).delete(synchronize_session=False)
+        log.info (f'Rows deleted-> {result}')
+        session.commit()
```
Comment (smarchint) [inline ext/scheduler/airflow/__lib.py:521]: use `airflow.utils.session.create_session`

File: ext/scheduler/airflow/__lib.py [hunk 2]
```diff
@@ -470,22 +469,25 @@ def optimus_sla_miss_notify(dag, task_list, blocking_task_list, slas, blocking_t
...
 
 def shouldSendSensorStartEvent(ctx):
     try:
+        from airflow import settings
         ti = ctx.get('task_instance')
         key = "sensorEvt/{}/{}/{}".format(ti.task_id , ctx.get('next_execution_date').strftime(TIMESTAMP_FORMAT) , ti.try_number)
-
-        result = ti.xcom_pull(key=key)
+        log.info( f'sensor calback check Xcom Key-> {key}')
+        session = settings.Session()
```
Comment (smarchint) [inline ext/scheduler/airflow/__lib.py:483]: use `airflow.utils.session.create_session` context manager

File: ext/scheduler/airflow/__lib.py [hunk 3]
```diff
@@ -470,22 +469,25 @@ def optimus_sla_miss_notify(dag, task_list, blocking_task_list, slas, blocking_t
...
     except Exception as e:
         print(e)
 
 def shouldSendSensorStartEvent(ctx):
     try:
+        from airflow import settings
         ti = ctx.get('task_instance')
         key = "sensorEvt/{}/{}/{}".format(ti.task_id , ctx.get('next_execution_date').strftime(TIMESTAMP_FORMAT) , ti.try_number)
-
-        result = ti.xcom_pull(key=key)
```
Comment (smarchint) [inline ext/scheduler/airflow/__lib.py:483]: need to understand why ti.xcom_pull did not work

Comment (smarchint) [reply ext/scheduler/airflow/__lib.py:483]: it it possible that before every ti schedule, xcom are deleted as per https://github.com/apache/airflow/blob/2.1.4/airflow/models/taskinstance.py#L1243

Comment (Mryashbhardwaj) [reply ext/scheduler/airflow/__lib.py:483]: image

you are right, it seems airflow is exhibiting strange inconsistent behaviour. This method has worked in past but dosent seem to be working now.
hence we are choosing to move to a diffrent approach

File: ext/scheduler/airflow/__lib.py [hunk 4]
```diff
@@ -470,23 +462,19 @@ def optimus_sla_miss_notify(dag, task_list, blocking_task_list, slas, blocking_t
...
 def shouldSendSensorStartEvent(ctx):
     try:
-        ti = ctx.get('task_instance')
-        key = "sensorEvt/{}/{}/{}".format(ti.task_id , ctx.get('next_execution_date').strftime(TIMESTAMP_FORMAT) , ti.try_number)
-
-        result = ti.xcom_pull(key=key)
-        if not result:
-            print("sending NEW sensor start event for attempt number ", ti.try_number)
-            ti.xcom_push(key=key, value=True)
+        ti=ctx['ti']
```
Comment (smarchint) [inline ext/scheduler/airflow/__lib.py:472]: unnecessary variable.

Comment (Mryashbhardwaj) [reply ext/scheduler/airflow/__lib.py:472]: this was referenced in 2 places , so i created a variable for clarity

**Review by smarchint — APPROVED** [review]
LGTM!

**Pull request merged**